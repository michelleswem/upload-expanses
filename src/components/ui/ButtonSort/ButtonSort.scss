@import '../../../assets/scss/variables';
.btn {
  cursor: pointer;
  background-color: transparent;
  border: none;
  vertical-align: middle;
  margin-left: 1rem;
}

.arrow-down {
  transform: rotate(180deg);
  transition: all 0.2s linear;
}

.arrow-up {
  transform: rotate(0);
  transition: all 0.2s linear;
}
