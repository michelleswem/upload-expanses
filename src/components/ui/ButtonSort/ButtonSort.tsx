import React from 'react';
import { Expanse } from '../../../models/expanse';
import './ButtonSort.scss';

type SortOrder = 'asc' | 'dec';
type Data = Expanse[];
type SortKey = keyof Data[0];

export type SortButtonProps = {
  onClick: () => void;
  className?: string;
  sortOrder: SortOrder;
  columnKey: SortKey;
  sortKey: SortKey;
  type?: 'button' | 'submit';
  children?: React.ReactNode;
};

export const ButtonSort: React.FC<SortButtonProps> = ({
  children,
  onClick,
  sortOrder,
  columnKey,
  sortKey,
}) => {
  return (
    <button
      onClick={onClick}
      className={`btn ${
        sortKey === columnKey && sortOrder === 'dec' ? 'arrow-down' : 'arrow-up'
      }`}
    >
      {children}
    </button>
  );
};
