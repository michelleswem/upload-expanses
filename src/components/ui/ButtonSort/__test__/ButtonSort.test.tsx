import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { ButtonSort, SortButtonProps } from '../ButtonSort';

const sortButtonProps: SortButtonProps = {
  onClick: jest.fn(),
  sortOrder: 'asc',
  columnKey: 'departments',
  sortKey: 'departments',
  type: 'button',
  children: 'button',
};

describe('Button component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('should render a button', () => {
    const { container } = render(<ButtonSort {...sortButtonProps} />);

    expect(container).toMatchSnapshot();
  });

  test('renders a button ', () => {
    //ARRANGE
    render(<ButtonSort {...sortButtonProps} type="button" />);

    // ACT
    // DO NOTHING

    // ASSERT
    const btnElement = screen.getByRole('button');
    expect(btnElement).toBeInTheDocument;
  });

  test('renders a button with a className that is Not', () => {
    // ARRANGE
    render(<ButtonSort {...sortButtonProps} />);

    // ASSERT
    expect(screen.getByRole('button')).not.toHaveClass('arrow-down');
  });

  test('renders a button in asc order with a className', () => {
    // ARRANGE
    render(
      <ButtonSort
        {...sortButtonProps}
        className="arrow-up"
        type="button"
        sortOrder="asc"
      />,
    );

    // ACT
    const buttonElement = screen.getByRole('button');
    userEvent.click(buttonElement);

    const outputElement = screen.getByRole('button');
    expect(outputElement).toHaveClass('arrow-up');
  });

  test('renders a button in dec order with a className', () => {
    // ARRANGE
    render(
      <ButtonSort
        {...sortButtonProps}
        className="arrow-down"
        type="button"
        sortOrder="dec"
        sortKey="departments"
        columnKey="departments"
      />,
    );

    // ACT
    const buttonElement2 = screen.getByRole('button');
    userEvent.click(buttonElement2);

    // ASSERT

    const outputElement2 = screen.getByRole('button');
    expect(outputElement2).toHaveClass('arrow-down');
  });

  test('should fire onClick callback', () => {
    render(
      <ButtonSort {...sortButtonProps} onClick={sortButtonProps.onClick} />,
    );

    const button = screen.getByRole('button');

    userEvent.click(button);

    expect(sortButtonProps.onClick).toHaveBeenCalledTimes(1);
  });
});
