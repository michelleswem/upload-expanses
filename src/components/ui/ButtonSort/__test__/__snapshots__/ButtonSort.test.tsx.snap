// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component renders a button 1`] = `
<div>
  <button
    class="btn arrow-up"
  >
    button
  </button>
</div>
`;
